ko/lessons/advanced/metaprogramming.md

```diff
@@ -0,0 +1,238 @@
+---
+layout: page
+title: 메타 프로그래밍
+category: advanced
+order: 7
+lang: ko
+---
+
+메타 프로그래밍은 코드를 사용해서 코드를 작성하는 방법입니다. 이를 통해 Elixir에서는 필요에 따라 언어를 확장할 수 있으며, 동적으로 코드를 변경할 수 도 있습니다. Elixir가 밑에서 어떤식으로 표현되고 있는지를 확인하는 부분부터 시작합시다. 그리고 어떻게 변경할 수 있는지, 마지막으로 이를 통해서 어떻게 확장할 수 있는지를 배워보겠습니다.
+
+주의: 메타 프로그래밍은 다루기 어려우며, 정말 필요할 때에만 사용해야 합니다. 과도한 사용은 이해하기도 어렵고 디버깅하기도 어려운 복잡한 코드를 만듭니다.
+
+{% include toc.html %}
+
+## Quote
+
+메타 프로그래밍의 첫번째 단계는 표현식이 어떻게 나타나는지를 이해하는 것입니다. Elixir 코드는 내부적으로 추상 문법 트리(Abstract Syntax Tree, AST)로 관리되며, 이는 튜플로 구성됩니다. 이 튜플들은 3개 정보를 포함합니다: 이는 각각 함수 이름, 메타 데이터, 함수의 인자들입니다.
+
+이러한 내부 구조를 확인하기 위해서 Elixir는 `quote/2` 함수를 제공합니다. `quote/2`를 사용해서 Elixir의 코드의 기저에 있는 표현식으로 변환할 수 있습니다:
+
+```elixir
+iex> quote do: 42
+42
+iex> quote do: "Hello"
+"Hello"
+iex> quote do: :world
+:world
+iex> quote do: 1 + 2
+{:+, [context: Elixir, import: Kernel], [1, 2]}
+iex> quote do: if value, do: "True", else: "False"
+{:if, [context: Elixir, import: Kernel],
+ [{:value, [], Elixir}, [do: "True", else: "False"]]}
+iex(6)>
+```
+
+처음의 3개는 튜플을 반환하지 않는 것을 확인하셨나요? 이 함수를 호출했을 때 자기 자신을 반환하는 리터럴이 5개 존재합니다.
+
+```elixir
+iex> :atom
+:atom
+iex> "string"
+"string"
+iex> 1 # All numbers
+1
+iex> [1, 2] # Lists
+[1, 2]
+iex> {"hello", :world} # 2 element tuples
+{"hello", :world}
+```
+
+## Unquote
+
+이제 코드의 내부 구조를 확인해 볼 수 있게 되었습니다만, 수정은 어떻게 할 수 있을까요? 새 코드나 값을 주입하기 위해서는 `unquote/1`를 사용합니다. `unquote/1`를 호출하면 호출된 값이 평가되어 AST에 주입됩니다. `unquote/1`의 동작을 확인하기 위해서 예를 몇 개 들어보죠:
+
+```elixir
+iex> denominator = 2
+2
+iex> quote do: divide(42, denominator)
+{:divide, [], [42, {:denominator, [], Elixir}]}
+iex> quote do: divide(42, unquote(denominator))
+{:divide, [], [42, 2]}
+```
+
+첫 번째 예제에서는 변수 `denominator`는 결과로 반환되는 AST가 변수에 접근하기 위한 튜플을 포함하게끔 만들어져 있습니다. `unquote/1` 예제에서 생성된 코드는 위에서 본 코드 대신 `denominator`의 값을 포함하고 있습니다.
+
+## 매크로
+
+`quote/2`와 `unquote/1`에 대해서 이해했다면, 이제 매크로에 뛰어들 시간입니다. 모든 메타 프로그래밍과 마찬가지로, 매크로는 정말 필요할 때에만 사용해야 합니다.
+
+가장 간단한 용어인 매크로는 애플리케이션 코드에 삽입할 수 있도록 감싸진 표현식을 반환하도록 설계된 특별한 함수입니다. 함수처럼 호출되는 것이 아니라 선언문을 감싸진 표현식으로 대체하는 모습을 상상해주세요. 매크로만 있다면 Elixir를 확장하거나, 애플리케이션에서 동적으로 코드를 추가할 수 있습니다.
```

Comment:
`Imagine the macro being replaced with the quoted expression rather than called like a function.`
This sentence did not mention about `선언문(declaration statement)`.

How about `함수처럼 호출되는 것이 아니라, 명령문을 감싸진 표현식으로 대체하는 모습을 상상해주세요`? 
Also this translation is not good;;